### ISSUE TITLE
Allow for external data input 

### ISSUE THREAD

**Issue by def-mycroft** [def-mycroft/pairs#8]
currently, only data gotten from `yfinance` can be used. will want to allow for arbitrary data input according to some format.

**Comment by GenusGeoff** [comment]
I use [Tiingo](https://www.tiingo.com) for a lot of my data needs. They're quite good and they have the appropriate licensing to not have to "end-around" Yahoo! Finance licensing.

An API key is free and a certain amount of data per month, quite liberal, is provided for free. I pay $10 / month for 100 GB of data and that data is served up via JSON or CSV.

I also [try to] contribute to the [tiingo-python](https://github.com/hydrosquall/tiingo-python) GitHub.

**Comment by def-mycroft** [comment]
> I use [Tiingo](https://www.tiingo.com) for a lot of my data needs. They're quite good and they have the appropriate licensing to not have to "end-around" Yahoo! Finance licensing.
> 
> An API key is free and a certain amount of data per month, quite liberal, is provided for free. I pay $10 / month for 100 GB of data and that data is served up via JSON or CSV.
> 
> I also [try to] contribute to the [tiingo-python](https://github.com/hydrosquall/tiingo-python) GitHub.
> 
> Some of the strengths are a very clean data set and the ability to obtain either End-of-Day data or intraday resample frequency of as low as 1-minute increments. If someone is super-zealous, a WebSockets connection can be turned on for tick-by-tick live data.

I'm glad you mentioned it, I'll add a separate issue to add Tiingo as a choice of data source.

**Comment by def-mycroft** [comment]
I didn't describe it very well before, wanted to add the ability to use a static csv file or something like that.

**Comment by def-mycroft** [comment]
added issue #9 @GenusGeoff

**Comment by GenusGeoff** [comment]
> I didn't describe it very well before, wanted to add the ability to use a static csv file or something like that.

Yes, I do understand that, as well. The simplest form would simply be to expect the data in a certain format. In fact, using the Tiingo data would be trivial if there was simply a standard input format. Alternately, the fields could simply be "mapped" on input.

Either way is fine, I was just suggesting it. I did finally find where all of the "working" code is in the repo. Until your Discord comments last night I kept checking into different files looking for the "meat" of the code unsuccessfully.

**Comment by def-mycroft** [comment]
that all makes sense. I want to add Tiingo as a data source, added #10 

yes, the main code is kind of buried in all of the code for `cement`. it should probably be in `pairs/core` instead of `pairs/lib`, that is maybe a bit more typical of python projects

**Comment by def-mycroft** [comment]
this issue needs #10 first, as does #9
